# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: netwerk/base/nsNetUtil.cpp
# Commit: 1d0091c56339
# Full Hash: 1d0091c5633982e358737a96ed8af1a2be00bf61
# Author: smayya <smayya@mozilla.com>
# Date: 2025-07-27 09:20:01
# Regressor Bug: 1960627
# File Overlap Count: 1
# Description:
#   Bug 1960627 - add LNAPermissionRequest class for dispatching permission prompt requests. r=necko-reviewers,emz,kershaw
#   
#   LNAPermissionRequest is used for dispatching localhost and local-network permission prompts to the user.
#   Additionally, this patch also implements the transaction retry mechanism based on prompt action by the user.
#   This patch also modifies existing unit tests to validate the transaction retry mechanism.
# ==============================================================================

diff -r 0981f147243d -r 1d0091c56339 netwerk/base/nsNetUtil.cpp
--- a/netwerk/base/nsNetUtil.cpp	Sat Jul 26 21:51:08 2025 +0000
+++ b/netwerk/base/nsNetUtil.cpp	Sat Jul 26 21:51:08 2025 +0000
@@ -4178,7 +4178,28 @@
   return NS_OK;
 }
 
-bool IsLocalNetworkAccess(
+bool IsLocalHostAccess(
+    const nsILoadInfo::IPAddressSpace aParentIPAddressSpace,
+    const nsILoadInfo::IPAddressSpace aTargetIPAddressSpace) {
+  // Determine if the request is moving to a more private address space
+  // i.e. Public -> LocalHost
+  // Private -> LocalHost
+
+  return ((aTargetIPAddressSpace == nsILoadInfo::IPAddressSpace::Local) &&
+          (aParentIPAddressSpace == nsILoadInfo::IPAddressSpace::Public ||
+           aParentIPAddressSpace == nsILoadInfo::IPAddressSpace::Private));
+}
+
+bool IsPrivateNetworkAccess(
+    const nsILoadInfo::IPAddressSpace aParentIPAddressSpace,
+    const nsILoadInfo::IPAddressSpace aTargetIPAddressSpace) {
+  // Determine if the request is moving from public to private address space
+
+  return ((aTargetIPAddressSpace == nsILoadInfo::IPAddressSpace::Private) &&
+          (aParentIPAddressSpace == nsILoadInfo::IPAddressSpace::Public));
+}
+
+bool IsLocalOrPrivateNetworkAccess(
     const nsILoadInfo::IPAddressSpace aParentIPAddressSpace,
     const nsILoadInfo::IPAddressSpace aTargetIPAddressSpace) {
   // Determine if the request is moving to a more private address space
@@ -4189,26 +4210,8 @@
   // for private network access
   // XXX (sunil) add link to LNA spec once it is published
 
-  if (aTargetIPAddressSpace == nsILoadInfo::IPAddressSpace::Public ||
-      aTargetIPAddressSpace == nsILoadInfo::IPAddressSpace::Unknown) {
-    return false;
-  }
-  // Check if this is an access to a local resource from Public or Private
-  // network
-  if ((aTargetIPAddressSpace == nsILoadInfo::IPAddressSpace::Local) &&
-      (aParentIPAddressSpace == nsILoadInfo::IPAddressSpace::Public ||
-       aParentIPAddressSpace == nsILoadInfo::IPAddressSpace::Private)) {
-    return true;
-  }
-
-  // Check if this is an access to a Private Network resource from a Public
-  // network
-  if ((aTargetIPAddressSpace == nsILoadInfo::IPAddressSpace::Private) &&
-      (aParentIPAddressSpace == nsILoadInfo::IPAddressSpace::Public)) {
-    return true;
-  }
-
-  return false;
+  return IsPrivateNetworkAccess(aParentIPAddressSpace, aTargetIPAddressSpace) ||
+         IsLocalHostAccess(aParentIPAddressSpace, aTargetIPAddressSpace);
 }
 }  // namespace net
 }  // namespace mozilla